### PR TITLE
Add method to check for unknown error

### PR DESCRIPTION
**Pull request by imurchie**

### CODE REVIEW

File: lib/protocol/errors.js [hunk 1]
```diff
@@ -742,7 +742,12 @@ for (let ErrorClass of _.values(errors)) {
   }
 }
 
-
+function isUnknownError (err) {
+  return !err.constructor.name ||
+         !_.keys(errors).find(function (error) {
```
Comment (mykola-mokhnach) [inline lib/protocol/errors.js]: you could probably use `_.some`

File: lib/protocol/protocol.js [hunk 2]
```diff
@@ -352,7 +354,7 @@ function buildHandler (app, method, path, spec, driver, isSessCmd) {
       if (isErrorType(err, errors.ProxyRequestError)) {
         protocolLogger.error(`Encountered internal error running command:  ${JSON.stringify(err)} ${err.stack}`);
         actualErr = err.getActualError();
-      } else if (!isW3C && (!(isErrorType(err, ProtocolError) || isErrorType(err, errors.BadParametersError)))) {
+      } else if (!isW3C && isUnknownError(err)) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: I still don't quite understand why `!isW3C` is necessary here

Comment (mykola-mokhnach) [reply lib/protocol/protocol.js]: perhaps @dpgraham can explain

File: lib/protocol/protocol.js [hunk 3]
```diff
@@ -352,7 +354,7 @@ function buildHandler (app, method, path, spec, driver, isSessCmd) {
...
         protocolLogger.error(`Encountered internal error running command: ${err.stacktrace || err.stack}`);
         actualErr = new errors.UnknownError(err);
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: can we now revert the change in `lib/jsonwp-proxy/proxy.js`? Perhaps the whole try/catch is not needed there now

File: lib/protocol/protocol.js [hunk 4]
```diff
@@ -352,7 +354,7 @@ function buildHandler (app, method, path, spec, driver, isSessCmd) {
       if (isErrorType(err, errors.ProxyRequestError)) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: maybe it might be handy to just use duck typing here and check if the err object has the corresponding method/property

File: lib/protocol/protocol.js (hunk 2 above)
Comment (imurchie) [reply lib/protocol/protocol.js]: I have no idea. It was there, so I'm not ready to remove.

File: lib/protocol/protocol.js (hunk 3 above)
Comment (imurchie) [reply lib/protocol/protocol.js]: It can't be reverted, or _every_ proxied error will turn into an unknown error (see https://github.com/appium/appium-base-driver/blob/master/lib/protocol/protocol.js#L452-L458). Removing the whole catch in https://github.com/appium/appium-base-driver/blob/master/lib/jsonwp-proxy/proxy.js#L399-L405 might be possible.

Comment (mykola-mokhnach) [reply lib/protocol/protocol.js]: yes, lets remove the whole

File: lib/protocol/protocol.js (hunk 2 above)
Comment (dpgraham) [reply lib/protocol/protocol.js]: The `getResponseForW3CError` that's called afterwards covers the conversions so it was unnecessary.

Comment (mykola-mokhnach) [reply lib/protocol/protocol.js]: lets add an explanation comment here then

File: lib/protocol/errors.js [hunk 5]
```diff
@@ -742,7 +742,12 @@ for (let ErrorClass of _.values(errors)) {
   }
 }
 
-
+function isUnknownError (err) {
+  return !err.constructor.name ||
```
[the remark this answers is not in the document]
Comment (imurchie) [reply lib/protocol/errors.js]: It should _always_ be there (so should `name`, really). Even for bare objects:
```js
$ node
> a = {};
{}
> a.constructor
[Function: Object]
> a.constructor.name
'Object'
>
```
